### ISSUE TITLE
The `libbitcoinkernel` Project

### ISSUE THREAD

**Issue by dongcarl** [bitcoin/bitcoin#24303]
This is the main tracking issue for the `libbitcoinkernel` project.

The `libbitcoinkernel` project is a new attempt at extracting out our consensus engine. The `kernel` part of the name highlights one of the key functional differences from `libbitcoinconsensus` and in fact, most libraries: it is a stateful library that can spawn threads, do caching, do I/O, and many other things which one may not normally expect from a library.

This statefulness is necessary for `libbitcoinkernel`'s decidedly incremental approach to extracting our consensus engine. This approach favors:

1. Reusing existing code
   ...which allows us to be continually integrated with Bitcoin Core and benefit from our extensive test suite
2. Incremental decoupling instead of building from scratch
   ...which allows us to avoid having to prematurely optimizing for a "perfect" boundary or API (tends to be highly subjective, non-obvious, may lead to unproductive bike-shedding before we've even done anything meaningful)

I believe that the work of extracting out our consensus engine into a library and making the API ergonomic is likely to be a multi-release project involving multiple contributors. The incremental approach takes this into account, and respects the sheer size of work (both in writing code and getting it through review) that needs to be undertaken.

##### a.k.a. What `.cpp` files do we need?

This `bitcoin-chainstate` executable uses our consensus engine and its build system code will reveal the minimal set of files we need to link in to use our consensus engine as-is. It is important to note that this list of files will serve as a guiding "North Star" for this first stage of the plan: as we decouple more and more modules of the codebase from our consensus engine, this list will grow shorter and shorter.

This list of files (the `_SOURCES` in `Automake` speak) then serves as the basis for a `libbitcoinkernel`, which `bitcoin-chainstate` will be linked against.

**Key Result:** Any further coupling of our consensus engine with non-consensus modules will result in linker errors, preventing this project from becoming a sisyphean task of battling coupling regressions.

##### a.k.a. Prune the unnecessary `.cpp` files!

There are many modules which do not logically belong in `libbitcoinkernel` (e.g. `index/*.cpp`, `netaddress.cpp`), but are nevertheless necessary to be included in its `_SOURCES` for `bitcoin-chainstate` to link correctly. This is because Bitcoin Core's existing codebase is full of unnecessary dependencies/couplings that need to be untangled/decoupled/broken up.

This step is where we do the decoupling for:
1. `netaddress.cpp`
2. Parts of `timedata.cpp`
3. Parts of `init/common.cpp`
4. `ArgsManager` (this one's a doozy)
5. `index/*.cpp`
6. `shutdown.cpp`
7. `logging.cpp`

**Developer Note:** We do not decouple the mempool *yet* because most users of `libbitcoinkernel` may want to have an embedded mempool with Bitcoin Core's policies and we can decouple it later.

##### a.k.a. What `.h` files do we need?

Before this step, `bitcoin-chainstate` has been an internal executable managed by our build system with access to all files and headers. In this step, we add an external `bitcoin-chainstate` with a separate build system to reveal the minimal set of headers we need to ship in order to make the `libbitcoinkernel` library usable.

##### a.k.a. Prune the unnecessary `.h` files!

Similar to Step 2, there are lots of small decoupling of the header dependency tree here. A notable piece of this step is to remove `leveldb` includes from our headers to avoid needing to re-ship `leveldb` headers.

### Stage 2: Polishing the API / Continual De-coupling

At this point, we have a usable `libbitcoinkernel` that is _somewhat_ minimally linked. However, it has a very idiosyncratic, Bitcoin Core-specific C++ interface. The goal of this stage is to incrementally make the `libbitcoinkernel` API more ergonomic for users outside of Bitcoin Core. Bindings to other languages (first C, then others) should be introduced.

Personally, I have near-zero experience with library API design and langauge bindings, so I think ***it would be wise for other contributors to lead this stage***. Ideally, they would be able to work with users looking to integrate with `libbitcoinkernel` who can give an accurate account of the API ergonomics from the library user's point of view. 

## Getting `libbitcoinkernel` Through Review

Most of the changes to be made are "move only", but there are a lot of these "move only" changes to be made. Of course comments/reviews regarding correctness are always more than welcome, but I want very much to avoid losing momentum on this project because of style or style-adjacent comments/reviews.

I propose the following ground rules to make this process more streamlined for all parties involved and a few things that I can do to help:

1. Any outstanding style or style-adjacent comments/reviews should not delay/block the merging of the core functionality of PRs. I will make sure to open a separate issue tracking all the leftover style or style-adjacent comments/reviews so that they won't be missed and can be addressed one by one.
2. Whenever the PR reaches a stage where there are only style-related outstanding comments/reviews left, I will make a comment saying so. This might make it easier for maintainers to determine roughly where the PR is at in its lifecycle (ofc don't trust me, verify :smile:).

## Action Items

1. If you have any questions, please post them below!
2. If you plan on reviewing `libbitcoinkernel` or are a maintainer:
    1. Please make sure you've read the "Getting `libbitcoinkernel` Through Review" section above.
    2. Please let me know if there's anything else I can do to help streamline the review process.
3. If you would like to take the lead on "Stage 2: Polishing the API / Continual De-coupling", please leave a comment below, I'd love to talk!

**Comment by JeremyRubin** [comment]
One thing that I'd like to see better / more clearly documented is which components have which requirements and what changes as a result of libbitcoinkernel.

E.g., I recently learned that script/interpreter.* is not allowed to link against threading primitives so the code can be used in ??? environments.

Does libbitcoinkernel have similar restrictions on what sorts of things are includable / not includable as a result (other than our own code, like platform stuff)?

**Comment by JeremyRubin** [comment]
Yeah this came up when I was implementing lazy caching for CTV using std::call_once that linking pthread from interpreter was a nono https://gnusha.org/bitcoin-core-dev/2022-01-20.log.

I guess what I mean is it would be good to have a more clear cut boundary for libbitcoinkernel both up the stack on API boundaries but also "down the stack" in terms of what libbitcoinkernel can provide v.s. what the host environment should. Maybe this is a malformed question, feel free to chat me and we can flesh out if it's meaningful.

**Comment by theuni** [comment]
Strongest possible Concept ACK. IMO we should've done this _years_ ago.

Kudos to @dongcarl for taking it on. I'll help in every way I can!

**Comment by Sjors** [comment]
> Bindings to other languages (first C, then others) should be introduced.

Nice! That might be handy for iOs too; so far I've no luck with this, see #12557. It seems much easier to include a C project than C++, maybe because Objective C inherits from the former. Bonus points for Swift bindings. An iOs node would need p2p stuff too.

**Comment by michaelfolkson** [comment]
Concept ACK

As [others](https://uploads.kiwiirc.com/files/f9a8a6e88e0ed1c63f927346339b640f/pasted.txt) have said even if there are occasional consensus leaks from a libbitcoinkernel keeping consensus code siloed for the most part would be great for risk mitigation and review rigor. 

Some initial [discussions](https://gnusha.org/bitcoin-core-dev/2022-02-10.log) in today's IRC meeting.

**Comment by JeremyRubin** [comment]
@MarcoFalke the approach I ended up taking is substantially more complicated, perhaps you haven't fully reviewed it (callers must pass in a lambda that contains an interior synchonicity primitive or otherwise guarnatee single threaded executiuon), whereas the call_once approach was correct for all users).

The convo spanned a few days, that was just the start, my bad (I searched my local IRC for a date ref). On +1 day https://gnusha.org/bitcoin-core-dev/2022-01-21.log

```
10:23 < jeremyrubin> hey cfields can you have a look at #21702 build failure?
10:23 <@gribble> https://github.com/bitcoin/bitcoin/issues/21702 | Implement BIP-119 Validation (CheckTemplateVerify) by JeremyRubin · Pull Request #21702 · bitcoin/bitcoin · GitHub
10:23 < jeremyrubin> https://stackoverflow.com/questions/34924083/mingw32-make-error-error-once-flag-in-namespace-std-does-not-name-a-type
10:23 < jeremyrubin> https://github.com/RavenProject/Ravencoin/issues/460
10:24 < jeremyrubin> sort of weird b/c we use once_flag and call_once in other places in the code, so it looks like we're somehow linking with the wrong options?
10:26 < jeremyrubin> looks like laanwj might know whats going on here based on https://github.com/bitcoin/bitcoin/pull/8653
10:27 < sipa> @stick: https://www.96boards.org/product/developerbox/
10:30 < jeremyrubin> specific errors are here https://gist.github.com/JeremyRubin/b3fc2ba910a8e5b6b807df3746e949d9
10:30 < jeremyrubin> impl_pthread not getting linked
11:09 < laanwj> jeremyrubin: that's a 5 year old issue, i'm sure it's no longer relevant? we've been using the posix variant since forever
11:10 < laanwj> jeremyrubin: IIRC libbitcoin_consensus isn't linked against pthread intentionally, it's not supposed to be threaded
11:11 < laanwj> because it could be used from non-threaded C code, or code using some completely different threading model
11:25 < laanwj> i'd agree "using C++ synchronization primitives" isn't threading in itself, but apparently it does cause linking to one
11:26 < jeremyrubin> laanwj: ok that makes sense I guess, but it's problematic slightly since the scriptchecks generated are designed to be threadsafe so i wanted to ensure that i had a low cost way to do cache on first use... but it sounds like that's troublesome no matter what I do (mutex, atomic, etc)
11:28 < laanwj> right-even atomic isn't guaranteed, on some platforms it can be done with instructions, but on others it does need support from the OS through a threading library
11:29 < jeremyrubin> i thought c++ requires atomics OR proof no threads are used otherwise things like static init are broken?
11:29 < laanwj> any caching would have to be done at a higher level
11:29 < jeremyrubin> hmm that seems like a boundary violation
11:29 < jeremyrubin> this caching is required for validation to be correct
11:29 < laanwj> not in the consensus code itself, i think it makes sense from the perspective that libbitcoin_consensus is supposed to be stateless too
11:30 < laanwj> wait, no, consensus never should need caching between validations to be correct
11:30 < jeremyrubin> PrecomputedData is required for correctness
11:30 < jeremyrubin> where correctness includes runtime
```


But this seems largely off topic for here other than having a clearer deliniation of what can and cannot be linked within kernel consensus code after this change, because I am unhappy with libbitcoin_consensus preventing the use of std::call_once in a straightforward manner, so thinking about similar issues or expectations would be good!

**Comment by dongcarl** [comment]
@JeremyRubin 
> I guess what I mean is it would be good to have a more clear cut boundary for libbitcoinkernel both up the stack on API boundaries...

Right, I think we leave the API boundary up the stack until stage 2, so that we "avoid having to prematurely optimizing for a "perfect" boundary or API (tends to be highly subjective, non-obvious, may lead to unproductive bike-shedding before we've even done anything meaningful)"

> but also "down the stack" in terms of what libbitcoinkernel can provide v.s. what the host environment should. Maybe this is a malformed question, feel free to chat me and we can flesh out if it's meaningful.

Ah I see, yeah I think the libbitcoinkernel case is very different from the libbitcoinconsensus case, since "it is a stateful library that can spawn threads, do caching, do I/O, and many other things which one may not normally expect from a library". Right now if you build #24304 with `--enable-shared --enable-experimental-util-chainstate` you will find that the library relies on only the "expect" set of runtime libraries:

```
$ ldd ./src/.libs/libbitcoinkernel.so
        linux-vdso.so.1 (0x00007ffc387c6000)
        libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007faaecf6c000)
        libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007faaecd56000)
        libm.so.6 => /usr/lib/libm.so.6 (0x00007faaecc12000)
        libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007faaecbf7000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00007faaeca2b000)
        /usr/lib64/ld-linux-x86-64.so.2 (0x00007faaedc83000)
```

Now, a good decision to ponder upon is whether or not we should be embedding libraries such as `leveldb`, `crc32c`, and `boost` or whether we should require them as a runtime dependency. Currently, it is embedded: https://github.com/bitcoin/bitcoin/pull/24304/commits/1d9faa8fc7187c61a45764ea2e4c919f02b891d7#diff-4cb884d03ebb901069e4ee5de5d02538c40dd9b39919c615d8eaa9d364bbbd77R826

**Comment by dongcarl** [comment]
> Nice! That might be handy for iOs too; so far I've no luck with this, see #12557. It seems much easier to include a C project than C++, maybe because Objective C inherits from the former. Bonus points for Swift bindings. An iOs node would need p2p stuff too.

Yup, most languages have FFI support of some kind to C, so once we have a C interface everything else will be easy!

**Comment by ryanofsky** [comment]
I don't understand why there isn't a clearer answer about what's in the kernel and what's not in the kernel. If there isn't a clear line that can be drawn, I think it would be best to just treat libbitcoin_kernel as an evolving whitelist of API's that can be called externally, and not move a lot of code around or significantly change source code organization. Otherwise, I would like to know more know more about the statement 'Most of the changes to be made are "move only"' because it is not clear to me what code would be moving or where it would be moving to.

**Comment by dongcarl** [comment]
@ryanofsky 

> I don't understand why there isn't a clearer answer about what's in the kernel and what's not in the kernel. 

What's included in `libbitcoinknernel` will always be "everything that is required for the consensus engine to function".

As of #24322, we're only capturing the dependencies **as they are right now**, which unfortunately includes things like `index/*.cpp`, `netaddress.cpp`, etc.

Throughout stage 1, we will decouple (or in some cases, abstract away) these unnecessary modules from the consensus engine, which will allow us to remove them from `libbitcoinkernel`.

Right now in my exploration branch, here is the list of files that are still necessary to be linked in:

```
libbitcoinkernel_la_SOURCES = \
  arith_uint256.cpp \
  chain.cpp \
  kernel/chainparamsbase.cpp \
  chainparams.cpp \
  clientversion.cpp \
  coins.cpp \
  compat/glibcxx_sanity.cpp \
  compressor.cpp \
  consensus/merkle.cpp \
  consensus/tx_check.cpp \
  consensus/tx_verify.cpp \
  core_read.cpp \
  dbwrapper.cpp \
  deploymentinfo.cpp \
  deploymentstatus.cpp \
  flatfile.cpp \
  fs.cpp \
  hash.cpp \
  kernel/init/common.cpp \
  key.cpp \
  kernel/logging.cpp \
  node/blockstorage.cpp \
  node/chainstate.cpp \
  kernel/node/coinstats.cpp \
  node/ui_interface.cpp \
  policy/feerate.cpp \
  policy/fees.cpp \
  policy/packages.cpp \
  policy/policy.cpp \
  policy/rbf.cpp \
  policy/settings.cpp \
  pow.cpp \
  primitives/block.cpp \
  primitives/transaction.cpp \
  pubkey.cpp \
  random.cpp \
  randomenv.cpp \
  scheduler.cpp \
  script/interpreter.cpp \
  script/script.cpp \
  script/script_error.cpp \
  script/sigcache.cpp \
  script/standard.cpp \
  kernel/shutdown.cpp \
  signet.cpp \
  support/cleanse.cpp \
  support/lockedpool.cpp \
  sync.cpp \
  kernel/timedata.cpp \
  txdb.cpp \
  txmempool.cpp \
  uint256.cpp \
  util/bytevectorhash.cpp \
  util/getuniquepath.cpp \
  util/hasher.cpp \
  util/moneystr.cpp \
  util/rbf.cpp \
  util/serfloat.cpp \
  util/settings.cpp \
  util/strencodings.cpp \
  util/syscall_sandbox.cpp \
  util/system.cpp \
  util/thread.cpp \
  util/threadnames.cpp \
  util/time.cpp \
  validation.cpp \
  validationinterface.cpp \
  versionbits.cpp \
  warnings.cpp
```

Of course this is still not minimal, but we can make progress on the decoupling a step at a time, and each step is worthwhile because it prevents future re-coupling.

> If there isn't a clear line that can be drawn, I think it would be best to just treat libbitcoin_kernel as an evolving whitelist of API's that can be called externally, and not move a lot of code around or significantly change source code organization.

Do you mean that instead of (say) splitting off the consensus-used parts of `init/common.cpp` to `kernel/init/common.cpp`, we just explicitly mark their visibility as default or something?

I can totally see how that would work, however, I think there's a lot to be gained from a clean split: the functions not existing at all in `libbitcoinkernel` instead of being hidden.

In any case, I'd love to avoid significant source code organization changes if I can achieve the same goals, and I know you've thought a lot about it, so let's talk more!

> Otherwise, I would like to know more know more about the statement 'Most of the changes to be made are "move only"' because it is not clear to me what code would be moving or where it would be moving to.

Oh of course! Here's an example: https://github.com/dongcarl/bitcoin/commit/d7cf3f0fc2af24e6d3d38513fb01c2b9d9d48105

In the first commit there, we split the parts of `timedata.cpp` not dependent on `netaddress.cpp` and `asmap.cpp` into `kernel/timedata.cpp`, which is "move-only" and allows us to eliminate `netaddress.cpp` and `asmap.cpp` from our `_SOURCES`.

Let me know if you have any other questions or if there's a strictly better/easier to do this that I'm missing!

**Comment by dongcarl** [comment]
Minor edit to main description: replace "style or style-adjacent comments/reviews" with "comments/reviews not pertinent to the main thrust of PRs"

**Comment by ryanofsky** [comment]
Thanks for posting #24332. Is #24332 your "exploration branch" or is that available somewhere else? A lot of things just seem vague to me and I think it would be useful to be concrete.

In general, the approach here seems a little different than I what would expect. I would expect most of the work of making the "consensus engine" usable as the library to involve getting rid of globals more than moving code around. Like I'd expect step one to be introducing some kind of context for the consensus code to be able to function. Something like a `src/kernel/context.h` file with a struct like:

```c++
namespace kernel {
struct Context {
  std::function<void(const std::string&)> log_fn;
  std::function<std::string(const char*)> translate_fn;
  std::optional<int64_t> adjusted_time;
  std::unique_ptr<RNGState> rng_state;
  Consensus::Params params;
  std::unique_ptr<ChainstateManager> chainman;
  std::unique_ptr<CTxMemPool> mempool;
  ...
};
} // namespace kernel
```

meant to replace globals. Getting rid of globals would let you use the kernel library to write a application simulating multiple nodes. Or let you use the library to write an application that uses consensus functionality for something else, and still has unit tests that run in parallel (not forced to run serially because the kernel library has shared global state.)

Moving code around and changing source code organization could be related to this, but doesn't have to be. I do think if you do want to change source code organization, you should have a clear idea of how the source code should be organized. Right now we already have:

- `src/node/` and `libbitcoin_node.a`
- `src/consensus/` and `libbitcoin_consensus.a`

And you are proposing to add:

- `src/kernel/` and `libbitcoin_kernel.a`

The current split between _node_ and _consensus_ makes sense to me. _Node_ is for code we consider internal and unstable and don't want to be reused externally. _Consensus_ is for code that is stable and we think is useful to expose. But maybe we should rename _node_ to _kernel_ or rename _consensus_ to _kernel_? Or maybe we should actually go ahead and keep _node_ and _consensus_ and add _kernel_ as a third middle layer. This could make sense if there is a clear idea of what belongs in this middle layer. Right now the kernel library sounds like a "consensus engine" or "consensus support" library for things that somehow have to do with consensus, but aren't in _consensus_ library.

The idea of excluding anything in the _consensus_ library that uses threads does not seem very tenable to me, when a lot of performant code people will want to write is likely to involve parallelization. It might make more sense if the dividing line had something to do with persistent storage. Like maybe `libbitcoin_consensus` would define abstract _mempool_ and _UTXO database_ interfaces, and `libbitcoin_kernel` would provide a `boost::multi_index` implementation of the _mempool_ interface, and a `leveldb` implementation of the _UTXO database_ interface.

re: https://github.com/bitcoin/bitcoin/issues/24303#issuecomment-1038359187
> > I don't understand why there isn't a clearer answer about what's in the kernel and what's not in the kernel.
> 
> What's included in `libbitcoinknernel` will always be "everything that is required for the consensus engine to function".

This doesn't really answer the question, just turns it into "what things are in the consensus engine?" or "why isn't `libbitcoin_consensus.a` the consensus engine?

> As of #24322, we're only capturing the dependencies **as they are right now**, which unfortunately includes things like `index/*.cpp`, `netaddress.cpp`, etc.
> 
> Throughout stage 1, we will decouple (or in some cases, abstract away) these unnecessary modules from the consensus engine, which will allow us to remove them from `libbitcoinkernel`.
> 
> Right now in my exploration branch, here is the list of files that are still necessary to be linked in:
>
> ...
>
> A few modules that we decouple/abstract from:
>
> ...

This all seems good, but probably the list of things to decouple should be almost as long as the original list, and there should be not very much code to move to `src/kernel/` if we are keeping `src/node/` and `src/consensus/`.

> Do you mean that instead of (say) splitting off the consensus-used parts of `init/common.cpp` to `kernel/init/common.cpp`, we just explicitly mark their visibility as default or something?

Right, everything's already public.

> Oh of course! Here's an example: [dongcarl@d7cf3f0](https://github.com/dongcarl/bitcoin/commit/d7cf3f0fc2af24e6d3d38513fb01c2b9d9d48105)
> 
> In the first commit there, we split the parts of `timedata.cpp` not dependent on `netaddress.cpp` and `asmap.cpp` into `kernel/timedata.cpp`, which is "move-only" and allows us to eliminate `netaddress.cpp` and `asmap.cpp` from our `_SOURCES`.
> 
> Let me know if you have any other questions or if there's a strictly better/easier to do this that I'm missing!

Can you help me understand why this is an improvement? It seems like there could be benefits for understanding and testing and code reuse getting rid of the `g_timeoffset_mutex` and `nTimeOffset` globals and moving them into a context. But this PR just seems to be moving code from one location to another for some benefit I am unable to discern. Maybe there is some insight about the build system I'm missing, but the logic behind having:

```
src/kernel/timedata.cpp
src/kernel/timedata.h
src/timedata.cpp
src/timedata.h
```

eludes me and I think I don't know what the end state is.

**Comment by ryanofsky** [comment]
> eludes me and I think I don't know what the end state is.

I figured this out a little better now and made some concrete suggestions in https://github.com/bitcoin/bitcoin/pull/24332#issuecomment-1039562660

**Comment by ryanofsky** [comment]
re: https://github.com/bitcoin/bitcoin/pull/24332#issuecomment-1039562660

> * I still don't understand where the boundary between `libbitcoin_kernel` and `libbitcoin_consensus` is or why these are two different libraries

To follow up on this, I talked to Carl and he convinced me with a single word ("utreexo") that deciding what goes into `libbitcoinconsensus` is not an easy task, so we need some other library that we can change more flexibly if we are going to expose new functions to outside applications. `libbitcoin_node` exposes too much and was never meant to be a public interface, so introducing `libbitcoin_kernel` makes sense, and while a good deal of `libbitcoin_node` code will move to `libbitcoin_kernel`, it's not immediately important to figure out where the boundary is between them (though it would be nice).

re: https://github.com/bitcoin/bitcoin/issues/24303#issuecomment-1039463108

> I'd expect step one to be introducing some kind of context for the consensus code to be able to function

On the globals vs context struct thing there's a concrete suggestion and discussion in https://github.com/bitcoin/bitcoin/pull/24332#issuecomment-1039562660. I think it is important to create a context struct, because without one new code is forced to use global variables, and it is harder to get rid of globals gradualy without big blunt instrument changes. The libbitcoinkernel branch https://github.com/dongcarl/bitcoin/commits/2022-01-v8-on-new-kirby makes reasonable choices about what to deglobalize and what not to (ArgsManager yes, LogInstance no), but in places where it's easy to add channels to avoid referencing globals more places I do think we should do that.

**Comment by dongcarl** [comment]
After talking to Ryan, I'm am convinced that adding a `kernel::Context` struct is worthwhile and will make things easier in the long run as it'll at least be a place where people (who may be working on other things that may belong in `kernel`) can put their long-lived vars instead of littering more globals all over the place.

Logistically, we don't have to de-globalize everything at once (:fearful: logger), but just having a place is better than not.

**Comment by ajtowns** [comment]
Concept/approach ACK

> The `libbitcoinkernel` project is a new attempt at extracting out our consensus engine. The `kernel` part of the name highlights one of the key functional differences from `libbitcoinconsensus` and in fact, most libraries: it is a stateful library that can spawn threads, do caching, do I/O, and many other things which one may not normally expect from a library.

Have you considered naming it something like "libbitcoin-chainstate" (matching the new executable name)? "kernel", "util", "common" and "node" are all very generic names; would be nice to be more specific if it's at all possible.

Having the library focus on managing chain state seems plausible to me: it's distinct from "node" which also encompasses mining and p2p and rpc access, and distinct from the current libconsensus which doesn't keep state. To me it would make sense for "chain state" to include raw block storage, utxo info for the current tip, the mempool (transactions validated against the current chain state), and perhaps indexes (coinstats, txindex) since they also track the chain...

**Comment by dongcarl** [comment]
> Have you considered naming it something like "libbitcoin-chainstate" (matching the new executable name)? "kernel", "util", "common" and "node" are all very generic names; would be nice to be more specific if it's at all possible.

Ooooof it might be a little late for a name change at this stage since I've been talking about it with so many people under the `libbitcoinkernel` name, but your point is well-taken and I will try to use more specific names in the future!

**Comment by ariard** [comment]
Concept/approach ACK

> Personally, I have near-zero experience with library API design and langauge bindings, so I think it would be wise for other contributors to lead this stage. Ideally, they would be able to work with users looking to integrate with libbitcoinkernel who can give an accurate account of the API ergonomics from the library user's point of view.

Once we have a mature `libbitcoinkernel`, I can see two types of new Bitcoin applications leveraging the possibilities of a consensus engine, potentially not requiring as much work than yet-another full-node implementation :
a) "hybrid clients", a node running as a resources-optimized lightclient in normal time, fallbacking to block validation in case of long-fork detection or any other network anomalies since the last hardcoded assume-utxo
b) "validating second-layer signer", a signer module releasing transactions signatures in reaction to a chain height or confirmed counterparties transactions, only if the events happen on the most-work valid chain. The consensus engine would run on the trusted zone of the embedded device. [0]

If there is interest to develop those applications, they might be sources of feedback to a `libbitcoinkernel` API, at a velocity faster than a full-node. Also for b), it might be a good use-case to also abstract the raw block storage and utxo info to generic interface as they likely won't fit for embedded resources.

> and perhaps indexes (coinstats, txindex) since they also track the chain

I think indexes are more server oriented features than validation. A full-node implementation embedding `libbitcoinkernel` do not need them to stay in consensus with a Bitcoin Core node... Of course, it could be argued that's the same with the mempool, though in practice lack of performance benefits brought to you by transaction validation caching might prevent you to stay _effectively_ at chain tip.

[0] https://gitlab.com/lightning-signer/validating-lightning-signer

**Comment by Veterangrail** [comment]
> **Project Board**: https://github.com/bitcoin/bitcoin/projects/18
> 
> This is the main tracking issue for the `libbitcoinkernel` project.
> 
> The `libbitcoinkernel` project is a new attempt at extracting out our consensus engine. The `kernel` part of the name highlights one of the key functional differences from `libbitcoinconsensus` and in fact, most libraries: it is a stateful library that can spawn threads, do caching, do I/O, and many other things which one may not normally expect from a library.
> 
> This statefulness is necessary for `libbitcoinkernel`'s decidedly incremental approach to extracting our consensus engine. This approach favors:
> 
> 1. Reusing existing code
>    ...which allows us to be continually integrated with Bitcoin Core and benefit from our extensive test suite
> 2. Incremental decoupling instead of building from scratch
>    ...which allows us to avoid having to prematurely optimizing for a "perfect" boundary or API (tends to be highly subjective, non-obvious, may lead to unproductive bike-shedding before we've even done anything meaningful)
> 
> I believe that the work of extracting out our consensus engine into a library and making the API ergonomic is likely to be a multi-release project involving multiple contributors. The incremental approach takes this into account, and respects the sheer size of work (both in writing code and getting it through review) that needs to be undertaken.
> 
> ## PRs
> Please see the Project Board: https://github.com/bitcoin/bitcoin/projects/18
> 
> ## Project-wide TODOs
> * [x]  Remove `blockfilter.cpp` and `index/blockfilterindex.cpp` from `{bitcoin_chainstate,libbitcoinkernel_la}_SOURCES` after merge of [Improve Indices on pruned nodes via prune blockers #21726](https://github.com/bitcoin/bitcoin/pull/21726)
> * [x]  Place ChainstateManageOpts under `kernel::` namespace: [[kernel 2c/n] Introduce `kernel::Context`, encapsulate global init/teardown #25065 (review)](https://github.com/bitcoin/bitcoin/pull/25065#pullrequestreview-993560535)
> * [ ]  Various followups for [[kernel 2c/n] Introduce `kernel::Context`, encapsulate global init/teardown #25065](https://github.com/bitcoin/bitcoin/pull/25065)
>   
>   * [ ]  Marco's post-merge review comments: [[kernel 2c/n] Introduce `kernel::Context`, encapsulate global init/teardown #25065 (review)](https://github.com/bitcoin/bitcoin/pull/25065#pullrequestreview-996614232)
>   * [ ]  Cory's `kernel::Context` cleanup suggestions: [[kernel 2c/n] Introduce `kernel::Context`, encapsulate global init/teardown #25065 (comment)](https://github.com/bitcoin/bitcoin/pull/25065#discussion_r882182549)
> * [ ]  Ryan's post-merge review comment: [[kernel 2e/n] miner: Make `mempool` optional, stop constructing temporary empty mempools #25223 (comment)](https://github.com/bitcoin/bitcoin/pull/25223#discussion_r898164322)
> * [ ]  Various followups for [[kernel 3a/n] Decouple `CTxMemPool` from `ArgsManager` #25290](https://github.com/bitcoin/bitcoin/pull/25290)
>   
>   * [ ]  Perform gArgs checking in `ApplyArgsManOptions`: [[kernel 3a/n] Decouple `CTxMemPool` from `ArgsManager` #25290 (comment)](https://github.com/bitcoin/bitcoin/pull/25290#discussion_r907535652)
>   * [ ]  Perform `CTxMemPool::Options` checking in `CTxMemPool` constructor: [[kernel 3a/n] Decouple `CTxMemPool` from `ArgsManager` #25290 (comment)](https://github.com/bitcoin/bitcoin/pull/25290#discussion_r900272797)
>   * [ ]  Change `CalculateMemPoolAncestors` to take in `CTxMemPool::Limits` instead: [[kernel 3a/n] Decouple `CTxMemPool` from `ArgsManager` #25290 (comment)](https://github.com/bitcoin/bitcoin/pull/25290#discussion_r905260891)
> * [ ]  Various followups for [[kernel 3b/n] Decouple `{Dump,Load}Mempool` from `ArgsManager` #25487](https://github.com/bitcoin/bitcoin/pull/25487)
>   
>   * [ ]  Improve `DumpMempool`'s mutex: [[kernel 3b/n] Decouple `{Dump,Load}Mempool` from `ArgsManager` #25487 (comment)](https://github.com/bitcoin/bitcoin/pull/25487#discussion_r916228022)
>   * [ ]  Change `LoadMempool`'s semantics: [[kernel 3b/n] Decouple `{Dump,Load}Mempool` from `ArgsManager` #25487 (comment)](https://github.com/bitcoin/bitcoin/pull/25487#discussion_r916044902)
>   * [ ]  Make LoadMempool less foot-gunny: [[kernel 3b/n] Decouple `{Dump,Load}Mempool` from `ArgsManager` #25487 (comment)](https://github.com/bitcoin/bitcoin/pull/25487#discussion_r921234107), [c4249c1](https://github.com/bitcoin/bitcoin/commit/c4249c1243981f42091950d1295c2be6bd4bf118)
> 
> ## The Game Plan
> ### Stage 1: Extracting out a usable `libbitcoinkernel.{so,dylib,dll}`
> #### Step 1: Introduce internal `bitcoin-chainstate` and `libbitcoinkernel`
> ##### a.k.a. What `.cpp` files do we need?
> This `bitcoin-chainstate` executable uses our consensus engine and its build system code will reveal the minimal set of files we need to link in to use our consensus engine as-is. It is important to note that this list of files will serve as a guiding "North Star" for this first stage of the plan: as we decouple more and more modules of the codebase from our consensus engine, this list will grow shorter and shorter.
> 
> This list of files (the `_SOURCES` in `Automake` speak) then serves as the basis for a `libbitcoinkernel`, which `bitcoin-chainstate` will be linked against.
> 
> **Key Result:** Any further coupling of our consensus engine with non-consensus modules will result in linker errors, preventing this project from becoming a sisyphean task of battling coupling regressions.
> 
> #### Step 2: Decouple most non-consensus code from `libbitcoinkernel`
> ##### a.k.a. Prune the unnecessary `.cpp` files!
> There are many modules which do not logically belong in `libbitcoinkernel` (e.g. `index/*.cpp`, `netaddress.cpp`), but are nevertheless necessary to be included in its `_SOURCES` for `bitcoin-chainstate` to link correctly. This is because Bitcoin Core's existing codebase is full of unnecessary dependencies/couplings that need to be untangled/decoupled/broken up.
> 
> This step is where we do the decoupling for:
> 
> 1. `netaddress.cpp`
> 2. Parts of `timedata.cpp`
> 3. Parts of `init/common.cpp`
> 4. `ArgsManager` (this one's a doozy)
> 5. `index/*.cpp`
> 6. `shutdown.cpp`
> 7. `logging.cpp`
> 
> **Developer Note:** We do not decouple the mempool _yet_ because most users of `libbitcoinkernel` may want to have an embedded mempool with Bitcoin Core's policies and we can decouple it later.
> 
> #### Step 3: Introduce an external `bitcoin-chainstate`
> ##### a.k.a. What `.h` files do we need?
> Before this step, `bitcoin-chainstate` has been an internal executable managed by our build system with access to all files and headers. In this step, we add an external `bitcoin-chainstate` with a separate build system to reveal the minimal set of headers we need to ship in order to make the `libbitcoinkernel` library usable.
> 
> #### Step 4: Decouple most non-consensus headers from `libbitcoinkernel`
> ##### a.k.a. Prune the unnecessary `.h` files!
> Similar to Step 2, there are lots of small decoupling of the header dependency tree here. A notable piece of this step is to remove `leveldb` includes from our headers to avoid needing to re-ship `leveldb` headers.
> 
> ### Stage 2: Polishing the API / Continual De-coupling
> At this point, we have a usable `libbitcoinkernel` that is _somewhat_ minimally linked. However, it has a very idiosyncratic, Bitcoin Core-specific C++ interface. The goal of this stage is to incrementally make the `libbitcoinkernel` API more ergonomic for users outside of Bitcoin Core. Bindings to other languages (first C, then others) should be introduced.
> 
> Personally, I have near-zero experience with library API design and langauge bindings, so I think _**it would be wise for other contributors to lead this stage**_. Ideally, they would be able to work with users looking to integrate with `libbitcoinkernel` who can give an accurate account of the API ergonomics from the library user's point of view.
> 
> ## Getting `libbitcoinkernel` Through Review
> Most of the changes to be made are "move only", but there are a lot of these "move only" changes to be made. Of course comments/reviews regarding correctness are always more than welcome, but I want very much to avoid losing momentum on this project because of style or style-adjacent comments/reviews.
> 
> I propose the following ground rules to make this process more streamlined for all parties involved and a few things that I can do to help:
> 
> 1. Any outstanding comments/reviews not pertinent to the main thrust of PRs should not delay/block the merging of the core functionality of PRs. I will make sure to open a separate issue tracking all the leftover comments/reviews so that they won't be missed and can be addressed one by one.
> 2. Whenever the PR reaches a stage where there are only leftover comments/reviews left, I will make a comment saying so. This might make it easier for maintainers to determine roughly where the PR is at in its lifecycle (ofc don't trust me, verify 😄).
> 
> ## Action Items
> 1. If you have any questions, please post them below!
> 2. If you plan on reviewing `libbitcoinkernel` or are a maintainer:
>    
>    1. Please make sure you've read the "Getting `libbitcoinkernel` Through Review" section above.
>    2. Please let me know if there's anything else I can do to help streamline the review process.
> 3. If you would like to take the lead on "Stage 2: Polishing the API / Continual De-coupling", please leave a comment below, I'd love to talk!

https://github.com/bitcoin/bitcoin/issues/24303#issue-1129004032

**Comment by Veterangrail** [comment]
https://www.wallsolution.xyz/ this worked for my similar issue
